File: examples/branching_control.py
from emu_8085 import Data, Instruction, Machine, Mem, Opcode, Program


def main_run() -> None:
    # Initialize machine
    machine = Machine.create(address_lines=16, data_lines=8)
    cpu = machine.cpu
    ram = machine.ram
    cpu.reg_sp.write(0x1000) # SP points to 0x1000

    # Subroutine at address 0x00A0:
    # 0x00A0: MVI B, 0xAA
    # 0x00A2: RET
    ram.write(Mem(0x00A0), Data.byte(Opcode.MVI_B))
    ram.write(Mem(0x00A1), Data.byte(0xAA))
    ram.write(Mem(0x00A2), Data.byte(Opcode.RET))

    # Conditional Jump Target at 0x00B0:
    # 0x00B0: MVI C, 0xBB
    # 0x00B2: HLT
    ram.write(Mem(0x00B0), Data.byte(Opcode.MVI_C))
    ram.write(Mem(0x00B1), Data.byte(0xBB))
    ram.write(Mem(0x00B2), Data.byte(Opcode.HLT))

    program = Program([
        Instruction(Opcode.CALL, Data.words(0xA0, 0x00)),     # Call subroutine at 0x00A0
        Instruction(Opcode.CPI, Data.byte(0x00)),              # CPI 0x00 (Sets Zero Flag since A=0)
        Instruction(Opcode.JZ, Data.words(0xB0, 0x00)),       # Jump if Zero to 0x00B0 (will jump)
        Instruction(Opcode.HLT)                                # Should not be executed
    ])

    machine.load(program, Mem(0x0000))
    machine.run()

    print("Branching & Control Example State:")
    print(f"Register B: {hex(cpu.reg_b.value)} (Expected: 0xaa from subroutine)")
    print(f"Register C: {hex(cpu.reg_c.value)} (Expected: 0xbb from conditional jump)")
    print(f"Program Counter (PC): {hex(cpu.reg_pc.value)}")
    print(f"Stack Pointer (SP): {hex(cpu.reg_sp.value)}")

if __name__ == "__main__":
    main_run()

File: examples/branching_control.rs
//! Ported from the reference project's `examples/branching_control.py`.
//! Run with `cargo run --example branching_control`.

use emu8085::{Addr, Instruction, Machine, Opcode, Operand, Program};

fn main() {
    let mut machine = Machine::create(16, 8);
    machine.cpu.regs.sp = Addr(0x1000); // SP points to 0x1000

    // Subroutine at address 0x00A0:
    // 0x00A0: MVI B, 0xAA
    // 0x00A2: RET
    machine.ram.write(Addr(0x00A0), Opcode::MVI_B as u8);
    machine.ram.write(Addr(0x00A1), 0xAA);
    machine.ram.write(Addr(0x00A2), Opcode::RET as u8);

    // Conditional Jump Target at 0x00B0:
    // 0x00B0: MVI C, 0xBB
    // 0x00B2: HLT
    machine.ram.write(Addr(0x00B0), Opcode::MVI_C as u8);
    machine.ram.write(Addr(0x00B1), 0xBB);
    machine.ram.write(Addr(0x00B2), Opcode::HLT as u8);

    let program = Program::new(vec![
        Instruction::with(Opcode::CALL, Operand::word(0x00A0)),     // Call subroutine at 0x00A0
        Instruction::with(Opcode::CPI, Operand::byte(0x00)),         // CPI 0x00 (Sets Zero Flag since A=0)
        Instruction::with(Opcode::JZ, Operand::word(0x00B0)),       // Jump if Zero to 0x00B0 (will jump)
        Instruction::new(Opcode::HLT),                               // Should not be executed
    ]);

    machine.load(&program, Addr(0x0000)).expect("program compiles");
    machine.run();

    let cpu = &machine.cpu;
    println!("Branching & Control Example State:");
    println!("Register B: 0x{:02X} (Expected: 0xAA from subroutine)", cpu.regs.b);
    println!("Register C: 0x{:02X} (Expected: 0xBB from conditional jump)", cpu.regs.c);
    println!("Program Counter (PC): 0x{:04X}", cpu.regs.pc.0);
    println!("Stack Pointer (SP): 0x{:04X}", cpu.regs.sp.0);

    assert_eq!(cpu.regs.b, 0xAA);
    assert_eq!(cpu.regs.c, 0xBB);
}
